assets/data-port/stepper/index.jsx

```diff
@@ -0,0 +1,102 @@
+import classnames from 'classnames';
+
+/**
+ * Helper method to update the steps when moving to the next one.
+ *
+ * @param  {Array} steps  The current steps.
+ * @return {Array}        The steps after moving.
+ */
+const moveToNext = ( steps ) => {
+	const newSteps = [ ...steps ];
+
+	for ( let i = 0; i < newSteps.length; i++ ) {
+		if ( newSteps[ i ].isActive ) {
+			newSteps[ i ].isComplete = true;
```

Comment:
A good strategy is always to return the new complete object. In this case, it would not be a problem now because our values are only primitive types, but if we create a new level in the object in some moment, it could be a problem.

When we are cloning the object in the `...steps`, it clones the object, creating a new reference, but it'll use the reference of the props.

Here we could use something like (I'm sorry by repeating the validations and some error, but only for example):

```javascript
newSteps[ i ] = {
  ...newSteps[ i ]
  isComplete: true,
  isActive: ! ( i + 1 < newSteps.length )
};
newSteps[ i + 1 ] = {
  ...newSteps[ i + 1 ],
  isActive: i + 1 < newSteps.length
};
```

Also, I think in this case, as we are modifying different items in each interaction (what prevents us to use the map), we could use the `Array.reduce`. But would not be a problem to keep as a `for`

Reply:
> When we are cloning the object in the ...steps, it clones the object, creating a new reference, but it'll use the reference of the props.

Why do we need to create new objects for the elements of the array though? I think that React is doing a 'shallow' reference comparison and since we create a new array every time it would be enough for React to understand that there was an update in the state and rerender.

> we could use the Array.reduce

Not sure if its appropriate to use reduce here since the semantics of the method is to apply an accumulator on the array to produce a single result. In our case we read the array to create an other array with elements of the same type so it could cause confusion.

Reply:
> Why do we need to create new objects for the elements of the array though? I think that React is doing a 'shallow' reference comparison and since we create a new array every time it would be enough for React to understand that there was an update in the state and rerender.

Yes! You are correct! It's a shallow comparison. So because that is important to "recreate" all the tree for the prop that we are changing.

I'm sorry that I didn't use a good example. Let's imagine that our component will be something like:

```javascript
<div>
  {steps.map( step =>
    <Step step={step}>
  )}
</div>
```

For the React, the `step` in this case, will never update because of the shallow comparison. Even we are updating the props, the reference for each object of the array is the same.

An example to run in the console:
```javascript
const steps = [
    { active: false }
];

const newSteps = [ ...steps ];

console.log( steps[0].active === newSteps[0].active ); // true

newSteps[0].active = true;

console.log( steps[0].active ); // true
```

Just reinforcing, probably it would not happen to the current code, but it would be easy to create an issue while refactoring.